Objects/genobject.c

```diff
@@ -1161,7 +1163,47 @@ PyTypeObject _PyCoroWrapper_Type = {
 PyObject *
 PyCoro_New(PyFrameObject *f, PyObject *name, PyObject *qualname)
 {
-    return gen_new_with_qualname(&PyCoro_Type, f, name, qualname);
+    PyObject *coro = gen_new_with_qualname(&PyCoro_Type, f, name, qualname);
+    if (!coro)
+        return NULL;
+
+    PyThreadState *tstate = PyThreadState_GET();
+    int depth = tstate->coroutine_origin_tracking_depth;
+
+    if (depth == 0) {
+        ((PyCoroObject *)coro)->cr_origin = NULL;
+    } else {
+        PyObject *origin = PyList_New(depth);
+        /* Immediately pass ownership to coro, so on error paths we don't have
+           to worry about it separately. */
```

Comment:
The idea is neat, but it makes the code harder to read and it's not how we usually do it in CPython. While reviewing the code I've already added 2 comments asking you to add `Py_DECREF(origin)`.  Please do it the long way: create, populate, cleanup on errors, and finally assign it to `cr_origin`. After all it's just two places where an extra decref is needed.